### PR TITLE
RUN-1079 - adding flag groups to submit and submit-mpi commands

### PR DESCRIPTION
**Pull request by moshe-kabala**

### CODE REVIEW

File: cmd/flags/groups.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	fg, found := fgm.m[groupName]
+	if !found {
+		fg = NewFlagsGroup(groupName)
+		fgm.m[groupName] = fg
+		*fgm.order = append(*fgm.order, groupName)
+	}
+	return fg.FlagSet
+}
+
+func (fgm FlagGroupMap) Groups() []FlagGroup {
```
Comment (yodarshafrir1) [inline cmd/flags/groups.go:51]: method should be private

File: cmd/flags/groups.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+func NewFlagGroupMap(cmd *cobra.Command) FlagGroupMap {
+	return FlagGroupMap{
+		cmd:   cmd,
+		m:     map[string]FlagGroup{},
+		order: &([]string{}),
+	}
+}
+
+func (fg *FlagGroup) Usage() string {
```
Comment (yodarshafrir1) [inline cmd/flags/groups.go:30]: method should be private

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]
Please see my notes

File: cmd/job/submit/submit.go [hunk 3]
```diff
@@ -186,46 +196,57 @@ func (submitArgs *submitArgs) addCommonFlags(cmd *cobra.Command) {
...
-	cmd.Flags().StringVar(&(submitArgs.CPULimit), "cpu-limit", "", "CPU limit for the job (0.5, 1)")
-	cmd.Flags().StringVar(&(submitArgs.MemoryLimit), "memory-limit", "", "Memory limit for this job (1G, 20M)")
-	cmd.Flags().StringVarP(&(submitArgs.Project), "project", "p", "", "Specifies the project to which the command applies. By default, commands apply to the default project. To change the default project use 'runai project set <project name>'.")
-	cmd.Flags().StringVarP(&(submitArgs.User), "user", "u", defaultUser, "Use different user to run the Job.")
-	cmd.Flags().StringVarP(&(submitArgs.Image), "image", "i", "", "Container image to use when creating the job.")
-	cmd.Flags().StringArrayVar(&(submitArgs.Args), "args", []string{}, "Arguments to pass to the command run on container start. Use together with --command.")
-	cmd.Flags().StringVar(&(submitArgs.NodeType), "node-type", "", "Enforce node type affinity by setting a node-type label.")
-	cmd.Flags().StringArrayVarP(&(submitArgs.EnvironmentVariable), "environment", "e", []string{}, "Set environment variables in the container.")
-	cmd.Flags().MarkHidden("user")
+	fs := fgm.GetOrAddFlagSet(AliasesAndShortcuts)
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/job/submit/submit.go:199]: change to flagSet

File: cmd/flags/groups.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+type FlagGroup struct {
+	Title   string
+	FlagSet *pflag.FlagSet
+}
+
+type FlagGroupMap struct {
+	cmd   *cobra.Command
+	m     map[string]FlagGroup
+	order *[]string
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go:19]: change

File: cmd/flags/groups.go [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+func (fgm FlagGroupMap) Groups() []FlagGroup {
+	groups := []FlagGroup{}
+	for _, name := range *fgm.order {
+		groups = append(groups, fgm.m[name])
+	}
+	return groups
+}
+
+func (fgm FlagGroupMap) ConnectToCmd() {
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go:59]: It does not just add it also override the help template
maybe UpdateFlagsByGroupsToCmd ?

File: cmd/flags/groups.go [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+Flags:
+` + usage + `{{end}}{{if .HasAvailableInheritedFlags}}
+Global Flags:
+{{.InheritedFlags.FlagUsages | trimTrailingWhitespaces}}{{end}}{{if .HasHelpSubCommands}}
+Additional help topics:{{range .Commands}}{{if .IsAdditionalHelpTopicCommand}}
+{{rpad .CommandPath .CommandPathPadding}} {{.Short}}{{end}}{{end}}{{end}}{{if .HasAvailableSubCommands}}
+Use "{{.CommandPath}} [command] --help" for more information about a command.{{end}}`))
+}
+
+func FlagsGroupUsage(fgs ...FlagGroup) string {
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go:90]: done

File: cmd/job/submit/submit_mpi.go [hunk 7]
```diff
@@ -33,6 +34,10 @@ import (
 
 const (
 	SubmitMpiCommand = "submit-mpi"
+	mpiExamples = `
+runai submit-mpi dist1 --processes=2 -g 1 \
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/job/submit/submit_mpi.go:38]: done

File: cmd/flags/groups.go [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+type FlagGroup struct {
+	Title   string
+	FlagSet *pflag.FlagSet
+}
+
+type FlagGroupMap struct {
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go:16]: done

File: cmd/flags/groups.go (hunk 1 above)
Comment (moshe-kabala) [reply cmd/flags/groups.go:51]: done

File: cmd/flags/groups.go (hunk 2 above)
Comment (moshe-kabala) [reply cmd/flags/groups.go:30]: done

File: cmd/flags/groups.go [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+		groups = append(groups, fgm.m[name])
+	}
+	return groups
+}
+
+func (fgm FlagGroupMap) ConnectToCmd() {
+	ConnectFlagsGroupToCmd(fgm.cmd, fgm.Groups()...)
+}
+
+// ConnectFlagsGroupToCmd
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go:63]: done

File: cmd/flags/groups.go [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+func ConnectFlagsGroupToCmd(cmd *cobra.Command, fgs ...FlagGroup) {
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go:64]: done

File: cmd/flags/groups.go [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+type FlagGroup struct {
+	Title   string
+	FlagSet *pflag.FlagSet
+}
+
+type FlagGroupMap struct {
+	cmd   *cobra.Command
+	m     map[string]FlagGroup
```
[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go:18]: I use it

[the remark this answers is not in the document]
Comment (moshe-kabala) [reply cmd/flags/groups.go]: fixed

File: cmd/job/submit/submit.go [hunk 12]
```diff
@@ -38,6 +38,16 @@ import (
 
 const (
 	runaiNamespace = "runai"
+
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit.go:41]: Please make a type for the group names.
`Scheduling`, `Storage`,  `Network` can be confusing if you don't mention that this is for flag groups

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]
last comment

File: cmd/flags/groups.go (hunk 2 above)
Comment (moshe-kabala) [reply cmd/flags/groups.go]: done

File: cmd/job/submit/submit.go (hunk 12 above)
Comment (moshe-kabala) [reply cmd/job/submit/submit.go:41]: done

File: cmd/job/submit/submit.go [hunk 13]
```diff
@@ -40,14 +40,14 @@ const (
...
 	// groups
-	AliasesAndShortcuts = "Aliases/Shortcuts"
-	ContainerDefinition = "Container Definition"
-	ResourceAllocation  = "Resource Allocation"
-	Storage             = "Storage"
-	Network             = "Network"
-	JobLifecycle        = "Job Lifecycle"
-	AccessControl       = "Access Control"
-	Scheduling          = "Scheduling"
+	AliasesAndShortcuts flags.FlagGroupName = "Aliases/Shortcuts"
```
Comment (yodarshafrir1) [inline cmd/job/submit/submit.go:43]: Please move these flags to the `groups` file

**Review by yodarshafrir1 — CHANGES_REQUESTED** [review]

File: cmd/job/submit/submit.go (hunk 13 above)
Comment (moshe-kabala) [reply cmd/job/submit/submit.go:43]: they do not belong to the submit?

Comment (moshe-kabala) [reply cmd/job/submit/submit.go:43]: the group is for the general case and not specific to the submit case

Comment (moshe-kabala) [reply cmd/job/submit/submit.go:43]: I can create in submit file that call submit_groups.go. is it OK?

Comment (yodarshafrir1) [reply cmd/job/submit/submit.go:43]: If you leave them here please add prefix to all the flags suggesting these are flag group names

Comment (moshe-kabala) [reply cmd/job/submit/submit.go]: done

**Review by yodarshafrir1 — APPROVED** [review]
💯

**Pull request merged**